---
title: "InterTADs User Guide"
author:
  - name: Nikolaos Pechlivanis
    affiliation: Institute of Applied Biosciences, 
      Centre for Research and Technology Hellas, Thessaloniki, GR
    email: nikosp41@certh.gr
  - name: Maria Tsagiopoulou
    affiliation: Institute of Applied Biosciences, 
      Centre for Research and Technology Hellas, Thessaloniki, GR
  - name: Maria Christina Maniou
    affiliation: Institute of Applied Biosciences, 
      Centre for Research and Technology Hellas, Thessaloniki, GR
  - name: Aspasia Orfanou
    affiliation: Institute of Applied Biosciences, 
      Centre for Research and Technology Hellas, Thessaloniki, GR
  - name: Fotis E. Psomopoulos
    affiliation: Institute of Applied Biosciences, 
      Centre for Research and Technology Hellas, Thessaloniki, GR

date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('InterTADs')`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{data_integration User Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
<!-- h1, h2, h3, h4 { -->
<!--   color:#17247a; -->
<!-- } -->

strong {
    color:#eb6b1c;
}

</style>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  error = FALSE,
  warning = FALSE,
  message = FALSE,
  crop = NULL
)
```

# Introduction

`InterTADs` is an open-source tool written in R, for integrating multi-omics 
data (e.g. DNA methylation, expression, mutation) from the same physical source 
(e.g. patient) taking into account the chromatin configuration of the genome, 
i.e. the topologically associating domains (TADs).

To install `InterTADs` package enter the following commands in 
your `R` session:

```{r eval = FALSE}
devtools::install_github("BiodataAnalysisGroup/InterTADs")
```

## Running InterTADs

Once `InterTADs` is successfully installed, it can be loaded as follow:

```{r setup}
library(InterTADs)
```

`InterTADs` can be run by calling a group of functions according to the 
following pipeline:

### Step 1 Data integration
```{r eval=FALSE}
    result<- data_integration (
        counts_folder = system.file(
            "extdata", "Datasets", "counts", package = "InterTADs"
        ),
        counts_fls = NULL,
    
        freq_folder = system.file(
            "extdata", "Datasets", "freq", package = "InterTADs"
        ),
        freq_fls = NULL,
    
        mapping_file = system.file(
            "extdata", "Datasets", "meta-data.csv", package = "InterTADs"),
    
        tad_file =system.file(
            "extdata", "Datasets",
            "hglft_genome_2dab_ec1330.bed", package = "InterTADs"),
    
        tech = "hg19"
    )
```


During the data integration part different NGS datasets are
combined into the same data table alonng with the related input 
data for the topologically associating domains (TADs).

- The function returns a list of integratedTADtable and summary. 
IntegratedTADtable contains all data info about TADs and 
ummary explains types of different events (methylation, expression etc.).

### Step 2 (optional) Ensembl ids 

```{r eval=FALSE}
result_ensmbl <- ensembl_ids(
    integratedTADtable = result[[1]],
    expr_data = 3
)
```

Converts external gene IDs to ensembl IDs. 
In case the user has no gene IDs or do not wish to convert 
the provided gene IDs to ensembl IDs, is able to skip this function.
Note that the functional analysis will perform only with ensembl IDs.

- The function returns a table with matched ensembl_ids.

### Step 3 (optional) Prepare methylation 

```{r eval=FALSE}
methylo_result <- prepare_methylation_values(
    integratedTADtable = result[[1]],
    mapping_file = system.file(
        "extdata", "Datasets", "meta-data.csv", package =
        "InterTADs"),
    meth_data = 2
)
```

In case DNA methylation datasets are provided this function reverses 
methylation values (0% -> 100%, 100% -> 0%). The user can skip this 
function in case it is not suitable for the provided dataset.

- The function returns an integrated TAD data table with the reversed 
methylation values is returned.

### Step 4 evenDiff
```{r eval=FALSE}
result_evenDiff <- evenDiff(
    integratedTADtable = result_ensmbl,
    mapping_file = system.file(
        "extdata", "Datasets", "meta-data.csv", package = "InterTADs"
    ),
    names.meta = c('group'),
    adj.PVal = 0.01,
    log_thr = 4
)
```
evenDiff performs differential analysis of multi-omics events (expression, 
methylation etc.).

- It returns a list of data tables containing significant events along with a 
summary file.

### Step 5 TADiff
```{r eval=FALSE}
TADiff_result <- TADiff(
    integratedTADtable = result_ensmbl,
    mapping_file = system.file(
        "extdata", "Datasets", "meta-data.csv", package = "InterTADs"
    ),
    names.meta = c("group"),
    expr_data = 3,
    adj.PVal = 0.01,
    log_thr = 2,
    tad_event = 4,
    pval_thr = 0.01,
    freq_thr = 20,
    mean_logFC_thr = 2
)
```

TADiff performs differential analysis of events (expression, methylation etc.)
by taking into account the chromatin configuration of the genome,
i.e. the topologically associating domains (TADs).

- Returns a list of data tables containing significant TADs along with
a summary file.

### Step 6 Functional Analysis
```{r eval=FALSE}
func_anal_res <- functional_analysis(
    annotation_file = "/gencode.v19.annotation.gff3.gz",
    tech = "hg19",
    files_evenDiff = result_evenDiff[[1]],
    files_TADiff = TADiff_result[[1]],
    names.meta = c('group'),
    exp_parent = 3,
    dbs = c("GO_Molecular_Function_2018",
        "GO_Biological_Process_2018",
        "KEGG_2019_Human"),
    type = c("GO_MF", "GO_BP", "KEGG"),
    genes_cover = c(11459, 14433, 7802),
    p_adjust_method = "fdr",
    cut_off = 0.05,
    criterio = "adjusted p-value",
    min_genes = 3,
    system = "win",
    dir_name =  system.file("extdata", "Datasets", package="InterTADs"),
    output_folder = paste0("Outputs"))
```

Enrichment analysis with GO terms, KEGG pathways and motif enrichment
with TFs. Optional step to be performed on the output datasets
of the `evenDiff.R` and `TADiff.R` scripts.


# Session info {.unnumbered}

Here is the output of `sessionInfo()` on the system on which this document was
compiled running pandoc ``r rmarkdown::pandoc_version()``:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

